x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -647,48 +625,64 @@ public void setFrequency(TimeValue frequency) {
-                        exception.getCause());
+                QUERY_TRANSFORMER.fromMap(query);
+            } catch(Exception ex) {
+                String msg = Messages.getMessage(Messages.DATAFEED_CONFIG_QUERY_BAD_FORMAT, id);
+
+                if (ex.getCause() instanceof IllegalArgumentException) {
+                    ex = (Exception)ex.getCause();
+                }
+
+                if (lenient) {
+                    logger.warn(msg, ex);
                 } else {
-                    throw ExceptionsHelper.badRequestException(
-                        Messages.getMessage(Messages.DATAFEED_CONFIG_QUERY_BAD_FORMAT, id, exception.getMessage()), exception);
+                    throw ExceptionsHelper.badRequestException(msg, ex);
                 }
             }
         }
 
-        public void setQuery(Map<String, Object> query) {
-            this.query = ExceptionsHelper.requireNonNull(query, QUERY.getPreferredName());
-        }
-
+        // Kept for easier testing
         public void setParsedAggregations(AggregatorFactories.Builder aggregations) {
             try {
                 setAggregations(AGG_TRANSFORMER.toMap(aggregations));
-            } catch (IOException | XContentParseException exception) {
+            } catch (Exception exception) {
                 // Certain thrown exceptions wrap up the real Illegal argument making it hard to determine cause for the user
                 if (exception.getCause() instanceof IllegalArgumentException) {
-                    throw ExceptionsHelper.badRequestException(
-                        Messages.getMessage(Messages.DATAFEED_CONFIG_AGG_BAD_FORMAT,
-                            id,
-                            exception.getCause().getMessage()),
-                        exception.getCause());
-                } else {
-                    throw ExceptionsHelper.badRequestException(
-                        Messages.getMessage(Messages.DATAFEED_CONFIG_AGG_BAD_FORMAT, id, exception.getMessage()), exception);
+                    exception = (Exception)exception.getCause();
                 }
+                throw ExceptionsHelper.badRequestException(
+                    Messages.getMessage(Messages.DATAFEED_CONFIG_AGG_BAD_FORMAT, id), exception);
             }
         }
 
         void setAggregations(Map<String, Object> aggregations) {
+            setAggregations(aggregations, true);
+        }
+
+        void setAggregations(Map<String, Object> aggregations, boolean lenient) {
             this.aggregations = aggregations;
+            try {
+                AGG_TRANSFORMER.fromMap(aggregations);
+            } catch (Exception ex) {
+                String msg = Messages.getMessage(Messages.DATAFEED_CONFIG_AGG_BAD_FORMAT, id);
+
+                if (ex.getCause() instanceof IllegalArgumentException) {
+                    ex = (Exception)ex.getCause();
+                }
+
+                if (lenient) {
+                    logger.warn(msg, ex);
```

Comment:
I think that data frame transforms would log a deprecation warning here.  If that's correct then it would be best for datafeeds to do the same.  Please check with @hendrikmuhs for the relevant code to copy.

Reply:
That's correct, I assume the same happens here. If it helps I can move the 2 classes for testing deprecation logging into core (MockDeprecatedQueryBuilder, MockDeprecatedAggregationBuilder).

Reply:
So, the `fromMap` parser does hit the deprecation logger in addition to simply collecting the deprecations. So, I think this comment is already handled, unless my thinking is incorrect.